enterprise/api/v1/views/enterprise_subsidy_fulfillment.py

```diff
@@ -94,47 +95,62 @@ class EnterpriseSubsidyFulfillmentViewSet(EnterpriseWrapperApiViewSet):
         (Http403): If the requesting user does not have the appropriate permissions.
         (EnrollmentModificationException): If something goes wrong while updating the platform CourseEnrollment object.
     """
-
-    def get_subsidy_fulfillment_queryset(self):
+    def get_permission_required(self):
         """
-        Return the queryset for this view. Queries across subsidy types until it finds a match for the provided uuid.
-        Returns a 404 if no subsidy fulfillment record is found.
+        Return specific permission name based on the view being requested
         """
-        enterprise_customer_uuid = get_enterprise_customer_from_user_id(self.request.user.id)
-        fulfillment_source_uuid = self.kwargs.get('fulfillment_source_uuid')
+        if self.action == 'unenrolled':
+            return ['enterprise.can_manage_enterprise_fulfillments']
+        elif self.action == 'retrieve':
+            return ['enterprise.can_access_admin_dashboard']
+        elif self.action == 'cancel_enrollment':
+            return ['enterprise.can_enroll_learners']
+        return []
+
+    def get_permission_object(self):
+        """
+        Depending on the requested view, returns a record identifier to check perms against.
+        """
+        if self.request.user.is_staff:
+            return None
 
-        # Get learner credit enrollments under the supplied fulfillment source uuid.
-        learner_credit_enrollments = models.LearnerCreditEnterpriseCourseEnrollment.objects.filter(
-            uuid=fulfillment_source_uuid
-        )
+        if self.action in ('retrieve', 'cancel_enrollment'):
+            enrollment_record = self.requested_fulfillment_source.enterprise_course_enrollment
+            return str(enrollment_record.enterprise_customer_user.enterprise_customer.uuid)
+        else:
+            return None
 
-        # Filters to match fulfillment enrollments' and entitlements' enterprise customer uuid to the requesting
-        # user's enterprise customer uuid.
-        subsidy_fulfillment_filter = Q(
-            enterprise_course_enrollment__enterprise_customer_user__enterprise_customer__uuid=enterprise_customer_uuid
-        )
-        subsidy_fulfillment_filter |= Q(
-            enterprise_course_entitlement__enterprise_customer_user__enterprise_customer__uuid=enterprise_customer_uuid
-        )
```

Comment:
This was the bad part that broken authorization - most `FulfillmentSource` records don't have a related enrollment, and this whole clause evaluated to `true`.

Reply:
Why would it be common that a `LearnerCreditEnterpriseCourseEnrollment` would not have a related enrollment? Don't we typically create these records at the same time we create the enrollment too?

Reply:
whoops, typo in my comment (but not the code)! most records don't have a related *entitlement*. All of them should have related enrollments.